# Importing the necessary libraries

In [64]:
import pandas as pd
import numpy as np

**Question 1:**


**What's the version of NumPy that you installed?**

In [2]:
np.__version__

'1.20.1'

**Loading the data**

In [3]:
df= pd.read_csv('C:/Users/HP/Documents/workspace/Ml-opzoomcamp/data.csv')

In [4]:
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

**Question 2**


**How many records are in the dataset?**

In [65]:
df.shape[0]

11914

There are 11914 rows in the data.

**Structure of the Data**

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

In [7]:
df.columns

Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')

**Question 3:**

**Who are the most frequent car manufacturers (top-3) according to the dataset?**

In [9]:
cd = df['Make'].value_counts()
cd

Chevrolet        1123
Ford              881
Volkswagen        809
Toyota            746
Dodge             626
Nissan            558
GMC               515
Honda             449
Mazda             423
Cadillac          397
Mercedes-Benz     353
Suzuki            351
BMW               334
Infiniti          330
Audi              328
Hyundai           303
Volvo             281
Subaru            256
Acura             252
Kia               231
Mitsubishi        213
Lexus             202
Buick             196
Chrysler          187
Pontiac           186
Lincoln           164
Oldsmobile        150
Land Rover        143
Porsche           136
Saab              111
Aston Martin       93
Plymouth           82
Bentley            74
Ferrari            69
FIAT               62
Scion              60
Maserati           58
Lamborghini        52
Rolls-Royce        31
Lotus              29
Tesla              18
HUMMER             17
Maybach            16
McLaren             5
Alfa Romeo          5
Spyker    

In [10]:
cd.head(3)

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

The top(3) manufacturers are Chevrolet, Ford, Volkswagen.

**Question 4:**


**What's the number of unique Audi car models in the dataset?**

In [67]:
(df[df['Make']=='Audi']).shape

(328, 16)

In [20]:
df1=df['Model'].value_counts()
df1

Silverado 1500        156
Tundra                140
F-150                 126
Sierra 1500            90
Beetle Convertible     89
                     ... 
Precis                  1
GS F                    1
Zephyr                  1
Sierra C3               1
Pulsar                  1
Name: Model, Length: 915, dtype: int64

In [21]:
newdata1 = df[df['Make'] =='Audi' ]
newdata1

Make Model  Year             Engine Fuel Type  Engine HP  \
17     Audi   100  1992             regular unleaded      172.0   
18     Audi   100  1992             regular unleaded      172.0   
19     Audi   100  1992             regular unleaded      172.0   
20     Audi   100  1992             regular unleaded      172.0   
21     Audi   100  1992             regular unleaded      172.0   
...     ...   ...   ...                          ...        ...   
10901  Audi   TTS  2016  premium unleaded (required)      292.0   
10902  Audi   TTS  2017  premium unleaded (required)      292.0   
11178  Audi    V8  1992             regular unleaded      276.0   
11179  Audi    V8  1993             regular unleaded      276.0   
11180  Audi    V8  1994             regular unleaded      276.0   

       Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors  \
17                  6.0            MANUAL  front wheel drive              4.0   
18                  6.0            MANUAL  front wheel drive              4.0   
19                  6.0         AUTOMATIC    all wheel drive              4.0   
20                  6.0            MANUAL  front wheel drive              4.0   
21                  6.0            MANUAL    all wheel drive              4.0   
...                 ...               ...                ...              ...   
10901               4.0  AUTOMATED_MANUAL    all wheel drive              2.0   
10902               4.0  AUTOMATED_MANUAL    all wheel drive              2.0   
11178               8.0         AUTOMATIC    all wheel drive              4.0   
11179               8.0         AUTOMATIC    all wheel drive              4.0   
11180               8.0         AUTOMATIC    all wheel drive              4.0   

                             Market Category Vehicle Size Vehicle Style  \
17                                    Luxury      Midsize         Sedan   
18                                    Luxury      Midsize         Sedan   
19                                    Luxury      Midsize         Wagon   
20                                    Luxury      Midsize         Sedan   
21                                    Luxury      Midsize         Sedan   
...                                      ...          ...           ...   
10901       Factory Tuner,Luxury,Performance      Compact         Coupe   
10902  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
11178                     Luxury,Performance      Midsize         Sedan   
11179                     Luxury,Performance      Midsize         Sedan   
11180                     Luxury,Performance      Midsize         Sedan   

       highway MPG  city mpg  Popularity   MSRP  
17              24        17        3105   2000  
18              24        17        3105   2000  
19              20        16        3105   2000  
20              24        17        3105   2000  
21              21        16        3105   2000  
...            ...       ...         ...    ...  
10901           27        23        3105  51900  
10902           27        23        3105  52500  
11178           18        13        3105   2000  
11179           18        13        3105   2149  
11180           18        13        3105   2239  

[328 rows x 16 columns]

In [23]:
newdata1['Model'].nunique()

34

The number of Unique Car model is 34.

**Question 5**


**How many columns in the dataset have missing values?**

In [26]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

**five(5) columns in the dataset has missing values.**

**Question 6**


- **Find the median value of "Engine Cylinders" column in the dataset.**


- **Next, calculate the most frequent value of the same "Engine Cylinders".**


- **Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.**


- **Now, calculate the median value of "Engine Cylinders" once again.**


- **Has it changed?**

In [27]:
df['Engine Cylinders'].median()

6.0

In [28]:
df['Engine Cylinders'].value_counts()

4.0     4752
6.0     4489
8.0     2031
12.0     230
5.0      225
10.0      68
0.0       56
3.0       30
16.0       3
Name: Engine Cylinders, dtype: int64

In [29]:
df['Engine Cylinders'].mode()

0    4.0
dtype: float64

In [30]:
df['Engine Cylinders'].fillna(4.0, inplace=True)

In [31]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders        0
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [32]:
df['Engine Cylinders'].median()

6.0

No, it didn't change.

**Question 7**


- **Select all the "Lotus" cars from the dataset.**


- **Select only columns "Engine HP", "Engine Cylinders".**


- **Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).**


- **Get the underlying NumPy array. Let's call it X.**


- **Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.**


- **Invert XTX.**


- **Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].**


- **Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.**


- **What's the value of the first element of w?**

In [33]:
#Newdata frame where the make is lotus
newdata2 = df[df['Make'] =='Lotus' ]
newdata2

Make      Model  Year                Engine Fuel Type  Engine HP  \
3912  Lotus      Elise  2009     premium unleaded (required)      189.0   
3913  Lotus      Elise  2009     premium unleaded (required)      218.0   
3914  Lotus      Elise  2009     premium unleaded (required)      189.0   
3915  Lotus      Elise  2010     premium unleaded (required)      189.0   
3916  Lotus      Elise  2010     premium unleaded (required)      218.0   
3917  Lotus      Elise  2011     premium unleaded (required)      189.0   
3918  Lotus      Elise  2011     premium unleaded (required)      217.0   
3919  Lotus      Elise  2011     premium unleaded (required)      217.0   
4216  Lotus     Esprit  2002     premium unleaded (required)      350.0   
4217  Lotus     Esprit  2003     premium unleaded (required)      350.0   
4218  Lotus     Esprit  2004     premium unleaded (required)      350.0   
4257  Lotus  Evora 400  2017     premium unleaded (required)      400.0   
4258  Lotus  Evora 400  2017     premium unleaded (required)      400.0   
4259  Lotus      Evora  2011     premium unleaded (required)      276.0   
4260  Lotus      Evora  2011     premium unleaded (required)      276.0   
4261  Lotus      Evora  2013     premium unleaded (required)      276.0   
4262  Lotus      Evora  2013     premium unleaded (required)      345.0   
4263  Lotus      Evora  2013     premium unleaded (required)      345.0   
4264  Lotus      Evora  2013     premium unleaded (required)      276.0   
4265  Lotus      Evora  2014     premium unleaded (required)      345.0   
4266  Lotus      Evora  2014     premium unleaded (required)      276.0   
4267  Lotus      Evora  2014     premium unleaded (required)      345.0   
4268  Lotus      Evora  2014     premium unleaded (required)      276.0   
4292  Lotus      Exige  2009  premium unleaded (recommended)      257.0   
4293  Lotus      Exige  2009  premium unleaded (recommended)      240.0   
4294  Lotus      Exige  2010  premium unleaded (recommended)      240.0   
4295  Lotus      Exige  2011  premium unleaded (recommended)      240.0   
4296  Lotus      Exige  2011  premium unleaded (recommended)      257.0   
4297  Lotus      Exige  2011  premium unleaded (recommended)      257.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   
3917               4.0            MANUAL  rear wheel drive              2.0   
3918               4.0            MANUAL  rear wheel drive              2.0   
3919               4.0            MANUAL  rear wheel drive              2.0   
4216               8.0            MANUAL  rear wheel drive              2.0   
4217               8.0            MANUAL  rear wheel drive              2.0   
4218               8.0            MANUAL  rear wheel drive              2.0   
4257               6.0            MANUAL  rear wheel drive              2.0   
4258               6.0            MANUAL  rear wheel drive              2.0   
4259               6.0            MANUAL  rear wheel drive              2.0   
4260               6.0            MANUAL  rear wheel drive              2.0   
4261               6.0            MANUAL  rear wheel drive              2.0   
4262               6.0            MANUAL  rear wheel drive              2.0   
4263               6.0            MANUAL  rear wheel drive              2.0   
4264               6.0            MANUAL  rear wheel drive              2.0   
4265               6.0            MANUAL  rear wheel drive              2.0   
4266               6.0            MANUAL  rear wheel drive              2.0   
4267              

In [34]:
newdata2.columns

Index(['Make', 'Model', 'Year', 'Engine Fuel Type', 'Engine HP',
       'Engine Cylinders', 'Transmission Type', 'Driven_Wheels',
       'Number of Doors', 'Market Category', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP'],
      dtype='object')

In [40]:
#selecting the columns Engine HP & Engine Cylinders
newdata3 =newdata2[['Engine HP','Engine Cylinders' ]]

In [45]:
#droping duplicates in the dataframe
newdata3 =newdata2[['Engine HP','Engine Cylinders' ]].drop_duplicates()

In [46]:
#shape of the new dataframe
newdata3.shape

(9, 2)

In [47]:
newdata3

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [48]:
newdata3.shape

(9, 2)

In [52]:
#converting the dataframe to an array.
X = np.array(newdata3)
X 

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [54]:
#Transposing the array
X_T = X.T
X_T

array([[189., 218., 217., 350., 400., 276., 345., 257., 240.],
       [  4.,   4.,   4.,   8.,   6.,   6.,   6.,   4.,   4.]])

In [55]:
#Multiplying the transposed array X by the array X (matrix ,matrix multiplication)
XTX = X_T.dot(X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [57]:
#Create an array Y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]
Y =np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
Y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

In [58]:
# Inverse of XTX
inv_XTX = np.linalg.inv(XTX)
inv_XTX

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [59]:
# Multiply the inverse of XTX with the transpose of X
XTX_1 = inv_XTX.dot(X_T)
XTX_1

array([[-1.31950096e-03,  2.84443321e-04,  2.29134897e-04,
        -4.18763778e-03,  4.46417989e-03, -2.39406462e-03,
         1.42221660e-03,  2.44147184e-03,  1.50122864e-03],
       [ 8.60893170e-02,  7.36567735e-04,  3.67976598e-03,
         2.54588185e-01, -2.13748621e-01,  1.51207962e-01,
        -5.18727169e-02, -1.14048164e-01, -6.40137937e-02]])

In [62]:
#Multipy XTX_1 by Y 
w = XTX_1.dot(Y)
w

array([  4.59494481, -63.56432501])

In [63]:
#The value of the first element
print("The Value of first element of w is: ",  w[0])


The Value of first element of w is:  4.594944810094579
